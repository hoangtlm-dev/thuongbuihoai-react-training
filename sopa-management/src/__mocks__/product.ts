const MOCK_PRODUCTS = [
  {
    name: 'Black shoes',
    price: 87,
    size: ['Small', 'Medium', 'Large'],
    description:
      'Our goal in designing the super-soft Atoms Crew Socks was to provide a seamless match with your Atoms shoes, enhancing your comfort for everyday wear - making them the perfect socks for your everyday shoes.',
    color: 'Black',
    image: {
      url: '/images/products/black-shoes.png',
      alt: 'This is black shoes'
    },
    id: '1'
  },
  {
    name: 'Blue shoes',
    price: 45,
    size: ['Small', 'Medium', 'Large'],
    description:
      'Our goal in designing the super-soft Atoms Crew Socks was to provide a seamless match with your Atoms shoes, enhancing your comfort for everyday wear - making them the perfect socks for your everyday shoes.',
    color: 'Blue',
    image: {
      url: '/images/products/blue-shoes.png',
      alt: 'This is blue shoes'
    },
    id: '2'
  }
];

const MOCK_PRODUCT = {
  name: 'Black shoes',
  price: 87,
  size: ['Small', 'Medium', 'Large'],
  description:
    'Our goal in designing the super-soft Atoms Crew Socks was to provide a seamless match with your Atoms shoes, enhancing your comfort for everyday wear - making them the perfect socks for your everyday shoes.',
  color: 'Black',
  image: {
    url: '/images/products/black-shoes.png',
    alt: 'This is black shoes'
  },
  id: '1'
};

export {
  MOCK_PRODUCTS,
  MOCK_PRODUCT
};
