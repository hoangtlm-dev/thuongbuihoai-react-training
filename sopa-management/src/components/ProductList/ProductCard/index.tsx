import { memo } from "react"
import Text, { SizeType, ThemeType } from "../../common/Text";
import Price, { PriceType } from "../../common/Price";
import Logo from "../../common/Logo";
import './productCard.css';

interface ProductCardProps {
  name?: string;
  color?: string;
  price?: number;
  href: string;
  src: string;
  alt?: string;
};

const ProductCard: React.FC<ProductCardProps> = ({
  name,
  color,
  price,
  href,
  src,
  alt
}) => {
  return (
    <div className="card">
      <Logo href={href} url={src} alt={alt} className="card-image" />
      <div className="card-item">
        <Text text={name} type={SizeType.regular} />
        <div className="card-info">
          <Text text={color} type={ThemeType.highlightPrimary} />
          <Price value={price} type={PriceType.primary} />
        </div>
      </div>
    </div>
  )
}

export default memo(ProductCard);