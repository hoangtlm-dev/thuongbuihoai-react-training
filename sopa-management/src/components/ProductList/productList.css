.product {
  padding: 80px 135px;
}

.product-list {
  display: flex;
  flex-wrap: wrap;
  margin: 0 auto;
  width: 1170px;
  justify-content: space-between;
}

.product-btn {
  text-align: center;
}