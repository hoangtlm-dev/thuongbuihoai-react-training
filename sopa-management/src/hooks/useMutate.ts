import { useMutation, useQueryClient } from 'react-query';

// Constants
import { QUERY_KEYS } from '@constants/keyQuery';
import { CART_URL } from '@constants/url';

// Services
import { api } from '@services/APIRequest';
import { Product } from '@interfaces/product';

/**
 * @description Custom hook post product
 */
export const useMutationPostProductToCart = () => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: (product: Product) => api.postData(CART_URL, product),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: [QUERY_KEYS.CARTS] });
    }
  });
};

/**
 * @description Custom hook edit product
 */
export const useMutationEditProductInCart = () => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: (product: Product) =>
      api.putData(`${CART_URL}/${product.id}`, product),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: [QUERY_KEYS.CARTS] });
    }
  });
};
