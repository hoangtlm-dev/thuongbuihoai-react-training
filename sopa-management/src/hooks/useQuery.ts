import { useInfiniteQuery, useQuery } from 'react-query';
import { AxiosError } from 'axios';


// Constants
import { QUERY_KEYS } from '@constants/keyQuery';
import { ACCOUNT_URL, PRODUCT_URL } from '@constants/url';
import {
  INITIAL_PRODUCT,
  LIMIT_PRODUCTS
} from '@constants/common';

// Interfaces
import { Product } from '@interfaces/product';

// Services
import { getData } from '@services/APIRequest';

// Stores
import { useProductStore } from '@stores/product';

// Helpers
import { flattenArray } from '@helpers/common';
import { Account } from '@interfaces/account';
import { useAccountStore } from '@stores/login';

/**
 * @description Fetch product by id
 * @param { string } id
 * @returns
 */
export const useFetchProductDetail = (id?: string) => {
  const { data, ...rest } = useQuery<Product, AxiosError>({
    queryKey: [QUERY_KEYS.PRODUCTS + id],
    queryFn: async () => await getData(`${PRODUCT_URL}/${id}`),
  });

  return {
    data: data || INITIAL_PRODUCT,
    ...rest
  };
};

/**
 * @description custom hook to get product with show more
 */
export const useInfiniteProducts = () => {
  const { setProducts } = useProductStore();

  const { data, ...rest } =  useInfiniteQuery<Product[], AxiosError>({
    queryKey: [QUERY_KEYS.PRODUCTS],
    queryFn: ({ pageParam = 1 }) => getData(PRODUCT_URL, pageParam),
    getNextPageParam: (lastPage, pages) => {
      const nextPage = pages.length + 1;
      return lastPage?.length > 0 && lastPage?.length === LIMIT_PRODUCTS
        ? nextPage
        : undefined;
    },
    onSuccess: ({ pages }) => {
      const result = flattenArray(pages);
      setProducts(result);
    }
  });

  return {
    data: data?.pages || [],
    ...rest
  }
};

/**
 * @description Fetch user
 * @returns
 */
export const useFetchUser = () => {
  const { setAccounts } = useAccountStore();

  return useQuery<Account[], AxiosError>({
    queryKey: [QUERY_KEYS.ACCOUNTS],
    queryFn: () => getData(ACCOUNT_URL),
    onSuccess: (data) => setAccounts(data),
  });
};
