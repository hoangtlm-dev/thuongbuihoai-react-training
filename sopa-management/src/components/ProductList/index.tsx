import { memo } from 'react';

// Hooks
import { useProducts } from '../../hooks/useQuery';

// Components
import ProductCard from './ProductCard';
import Button,
{
  ButtonType
} from '../common/Button';

// Constants
import { Product } from '../../interfaces/product';

// Styles
import './productList.css';

const ProductList: React.FC = () => {
  const {
    data,
    fetchNextPage,
    hasNextPage,
    isLoading,
    isFetchingNextPage
  } = useProducts();

  const isDisable = !hasNextPage || isFetchingNextPage;

  return (
    <div data-testId='product-list' className='product'>
      {isLoading ? (
        // TODO: I will create component Loading later
        <p>Loading...</p>
      ) : (
        <>
          <div className='product-list'>
            {data?.pages?.map((page) => (
              <>
                {page.map((product: Product) => (
                  <div key={product.id}>
                    <ProductCard
                      href={'/'}
                      src={product.image?.url}
                      name={product.name}
                      color={product.color}
                      price={product.price}
                    />
                  </div>
                ))}
              </>
            ))}
          </div>
          <div className='product-btn'>
            <Button
              children={
                isFetchingNextPage
                  ? 'Loading more...'
                  : hasNextPage
                  ? 'Show More'
                  : 'No more products'
              }
              type={ButtonType.primary}
              onClick={() => fetchNextPage()}
              disable={isDisable}
            />
          </div>
        </>
      )}
    </div>
  );
};

export default memo(ProductList);
