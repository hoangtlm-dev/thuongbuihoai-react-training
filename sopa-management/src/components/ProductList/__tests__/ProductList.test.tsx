// Libs
import '@testing-library/jest-dom';
import { render } from '@testing-library/react';

import ProductList from '..';
import { QueryClient, QueryClientProvider } from 'react-query';
import { MOCK_PRODUCTS } from '../../../__mocks__/product';
import { useProducts } from '../../../hooks/useQuery';

// Components
const queryClient = new QueryClient();
jest.mock('../../../hooks/useQuery');

const handleFetchNextPage = jest.fn();

(useProducts as jest.Mock).mockReturnValue({
  data: MOCK_PRODUCTS,
  isLoading: false,
  isValidating: false,
  error: '',
  fetchNextPage: handleFetchNextPage
});

describe('ProductList component', () => {
  test('should render ProductList component', () => {
    const { getByTestId } = render(
      <QueryClientProvider client={queryClient}>
        <ProductList />
      </QueryClientProvider>
    );
    expect(getByTestId('product-list')).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const { asFragment } = render(
      <QueryClientProvider client={queryClient}>
        <ProductList />
      </QueryClientProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
