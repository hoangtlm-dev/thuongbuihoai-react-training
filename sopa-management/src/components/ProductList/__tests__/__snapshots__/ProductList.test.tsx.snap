// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="product"
    data-testid="product-list"
  >
    <div
      class="product-list"
    />
    <div
      class="product-btn"
    >
      <button
        class=" btn btn-primary"
        data-testid="button"
        disabled=""
      >
        No more products
      </button>
    </div>
  </div>
</DocumentFragment>
`;
